### PR TITLE
Serializers for Workflow, Process and Step objects

### PR DESCRIPTION
**Pull request by styliii**
### Step
```
{
  "data":{
    "id":"1",
    "type":"step",
    "attributes":{
      "title":"file taxes",
      "completed":true,
      "kind":"default",
      "resource_url":null,
      "resource_title":null,
      "position":100
    }
  }
}
```

### Process
```

{
  "data": {
    "id": "1",
    "type": "process",
    "attributes": {
      "title": "first process/milestone",
      "effort": "small",
      "categories": [

      ],
      "status": "todo",
      "position": 100
    },
    "relationships": {
      "workflow": {
        "data": {
          "id": "2",
          "type": "workflow_instance_workflow"
        }
      },
      "steps": {
        "data": [
          {
            "id": "1",
            "type": "workflow_instance_step"
          }
        ]
      }
    }
  },
  "included": [
    {
      "id": "2",
      "type": "workflow",
      "attributes": {
        "name": "discovery",
        "description": "find out what wf is about"
      },
      "links": {
        "url": "/v1/workflows/2"
      }
    },
    {
      "id": "1",
      "type": "step",
      "attributes": {
        "title": "file taxes",
        "completed": true,
        "kind": "default",
        "resource_url": null,
        "resource_title": null,
        "position": 100
      }
    }
  ]
}
```

### Workflow
```

{
  "data": {
    "id": "2",
    "type": "workflow",
    "attributes": {
      "name": "discovery",
      "description": "find out what wf is about",
      "version": "1.0"
    },
    "relationships": {
      "processes": {
        "data": [
          {
            "id": "1",
            "type": "workflow_instance_process"
          }
        ]
      }
    },
    "links": {
      "url": "/v1/workflows/2"
    }
  },
  "included": [
    {
      "id": "1",
      "type": "process",
      "attributes": {
        "title": "first process/milestone",
        "effort": "small",
        "categories": [

        ],
        "status": "todo",
        "position": 100
      },
      "relationships": {
        "workflow": {
          "data": {
            "id": "2",
            "type": "workflow_instance_workflow"
          }
        },
        "steps": {
          "data": [
            {
              "id": "1",
              "type": "workflow_instance_step"
            }
          ]
        }
      }
    }
  ]
}
```

### CODE REVIEW

**Comment by styliii** [comment]
@keithtom should the serializer live in the workflow gem? I noticed in the config/routes you mounted the workflow gem engine.

**Comment by keithtom** [comment]
@styliii  serializer and all API code should not be in the gem but the core app.

File: app/serializers/v1/workflow/process_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+class ProcessSerializer
```
Comment (keithtom) [inline app/serializers/v1/workflow/process_serializer.rb:1]: if you look at all the other serializers, they inherit from `< ApplicationSerializer`.
That's going to put some conventions on your models that will make them error out.  Basically any model that goes into the API needs to have a non-incremental integer ID called `external_identifier`.  That's going to be a bit annoying though bc now we need to update the migrations and ` include ApplicationRecord::ExternalIdentifier` in the models.  See this model for example: https://github.com/keithtom/wildflower-platform/blob/main/app/models/address.rb

Also check out this migration I wrote for a similar prob:
https://github.com/keithtom/wildflower-platform/blob/main/db/migrate/20211117190245_add_external_identifiers.rb

don't need to include JSONAPI:Serializer either.

https://github.com/keithtom/wildflower-platform/blob/main/app/serializers/application_serializer.rb

File: app/serializers/v1/workflow/process_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    { name: object.workflow.name, description: object.workflow.description }
+  end
+
+  attribute :steps do |object|
+    object.steps.map do |step|
+      { title: step.title, completed: step.completed, kind: step.kind, resource_url: step.resource_url, position: step.position }
+    end
+  end
+end
+
```
Comment (keithtom) [inline app/serializers/v1/workflow/process_serializer.rb:16]: So the way I've been doing relationships is with the `has_many`
For example:
https://github.com/keithtom/wildflower-platform/blob/main/app/serializers/v1/school_serializer.rb
BTW, there's a bug so you have to write it like that even though it should be able to figure things out.  You can read more about it in this readme: https://github.com/keithtom/wildflower-platform/blob/main/app/serializers/README.md

It changes the JSON structure and is a bit more annoying for Taylor but I think this makes everything "standard JSON compliant API".   The con is that it isn't as convenient, the pro is that we have a standard way to interact w/ our api and we don't have to really think about it.  I kinda prefer just going w/ this method.

File: app/serializers/v1/workflow/workflow_serializer.rb [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
   include JSONAPI::Serializer
 
-  set_type :movie  # optional
-  set_id :owner_id # optional
-  attributes :name, :year
-  has_many :processes
-  belongs_to :owner, record_type: :user
-  belongs_to :movie_type
+  attributes :name, :description
+
```
Comment (keithtom) [inline app/serializers/v1/workflow/workflow_serializer.rb:5]: I think this should also have version attribute

File: app/serializers/v1/workflow/workflow_serializer.rb [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
+  attribute :processes do |object|
+    object.processes.map do |process|
+      { title: process.title, effort: process.effort, status: process.status, categories: process.categories }
+    end
+  end
+
+  link :url
```
Comment (keithtom) [inline app/serializers/v1/workflow/workflow_serializer.rb:12]: where does the value of URL come from?

File: app/serializers/v1/workflow/process_serializer.rb (hunk 2 above)
Comment (styliii) [reply app/serializers/v1/workflow/process_serializer.rb:16]: ok, i was worried about consistency as well

File: app/serializers/v1/workflow/workflow_serializer.rb (hunk 4 above)
Comment (styliii) [reply app/serializers/v1/workflow/workflow_serializer.rb:12]: i just made some dummy url function as a placeholder, until i figure out routes

**Comment by styliii** [comment]
@keithtom made the updates. Ready for re-review

**Pull request merged**